bindings/torch/tinycudann/modules.py

```diff
@@ -55,6 +55,79 @@ def backward(ctx, doutput):
 
 		return None, input_grad, weight_grad, None
 
+# NOTE: Jianfei: this is a temporary solution, when only grid has backward_backward and networks do not.
+class _module_function_encoding(torch.autograd.Function):
+	@staticmethod
+	def forward(ctx, native_tcnn_module, input, params, loss_scale):
+		# If no output gradient is provided, no need to
+		# automatically materialize it as torch.zeros.
+		ctx.set_materialize_grads(False)
+
+		native_ctx, output = native_tcnn_module.fwd(input, params)
+		ctx.save_for_backward(input, params, output)
+		ctx.native_tcnn_module = native_tcnn_module
+		ctx.native_ctx = native_ctx
+		ctx.loss_scale = loss_scale
+
+		return output
+
+	@staticmethod
+	def backward(ctx, doutput):
+		if doutput is None:
+			return None, None, None, None
+
+		if not doutput.is_cuda:
+			print("TCNN WARNING: doutput must be a CUDA tensor, but isn't. This indicates suboptimal performance.")
+			doutput = doutput.cuda()
+
+		input, params, output = ctx.saved_tensors
+		input_grad, weight_grad = _module_function_encoding_backward.apply(ctx, doutput, input, params, output)
+
+		return None, input_grad, weight_grad, None
+
+class _module_function_encoding_backward(torch.autograd.Function):
+	@staticmethod
+	def forward(ctx, ctx_fwd, doutput, input, params, output):
+		ctx.ctx_fwd = ctx_fwd
+		ctx.save_for_backward(input, params, doutput)
+		with torch.no_grad():
+			scaled_grad = doutput * ctx_fwd.loss_scale
+			input_grad, weight_grad = ctx_fwd.native_tcnn_module.bwd(ctx_fwd.native_ctx, input, params, output, scaled_grad)
+			input_grad = None if input_grad is None else (input_grad / ctx_fwd.loss_scale)
+			weight_grad = None if weight_grad is None else (weight_grad / ctx_fwd.loss_scale)
+		return input_grad, weight_grad
+
+	@staticmethod
+	def backward(ctx, dinput_grad, dweight_grad):
+		# NOTE: currently support:
+		#		✓	d(dL_dinput)_d(dL_doutput)	doutput_grad
+		#		✓	d(dL_dinput)_d(params)		weight_grad
+		#		x	d(dL_dinput)_d(input)
+		#		x	d(dL_dparam)_d(...)
+		# assert dweight_grad is None, "currently do not support 2nd-order gradients from gradient of grid"
+		input, params, doutput = ctx.saved_tensors
+		with torch.enable_grad():
+			# NOTE: preserves requires_grad info
+			# NOTE: be cautious when multiplying and dividing loss_scale:
+   			# dinput_grad = dinput_grad * ctx.ctx_fwd.loss_scale
+			doutput = doutput * ctx.ctx_fwd.loss_scale
```

Comment:
Is it important to do this in `with torch.enable_grad():` rather than just outside of `with torch.no_grad():`? Or in other words: if `doutput` doesn't require gradients in the first place, I don't think they need to be explicitly enabled.

Reply:
Yes, I found it necessary.

1. Shortly speaking, the backward.backward() function here is invoked implicitly in an `no_grad()` context by default. Hence, the product of multiplying a `requires_grad=True` Tensor (doutput) with a float (loss_scale), will be a `requires_grad=False` Tensor by default. The `requires_grad` property is lost by default.

Specifically speaking, consider the final `loss.backward()`, which typically will _NOT_ specify `create_graph=True` (and usually not any other parameters at all). Then, it will invoke all tensors `backwards` via a [run_backward()](https://github.com/pytorch/pytorch/blob/e813eaee22be2bfe5e6f2a137305cffe864a01da/torch/autograd/__init__.py#L173) funtion, which causing all the backwards() implicitly running in a `create_graph=False / no_grad()` context, including the one here. This introduces the above mentioned problem of not passing `requires_grad` property.

2. The underlying `bwd_bwd_input` in `bindings.cpp` relies on the `requires_grad()` property of `doutput` to behave corretly (to determine whether to compute `dL_ddLdoutput` or not). 

I've thought of other alternative ways, e.g. passing a bool `need_xxx_grad` from python to cpp,  or multiplying the `loss_scale` inside `bindings.cpp` instead of in `modules.py`. But, they may not be as elegant ? or as concise? comparing to the implementation now. So `requires_grad` property indeed matters.

Reply:
I see, that makes perfect sense -- I hadn't realized that the entire function is implicitly wrapped in `no_grad()`. Thank you for the clarification :)